### ISSUE TITLE
Filter for all answered topics by a user does not show all topics

### ISSUE THREAD

**Issue by MyLifeOrganized** [kinerity/bestanswer#2]
Steps: 
1) There is a user with 7 answered topics. 
2) Click the answered counter to see all these topics
Expected: All 7 topics listed in the search result
Actually: Only 3 topics are listed (the counter of the result still shows correct number though = 7):
http://www.mylifeorganized.net/forum-ru/search.php?author_id=2286&sr=topics&filter=topicsanswered
Note: if you are a registered user of the forum the search result shows only one topic with the same request.

Another example with bestanswer ext installed on different forum where more topics answered:
https://dmyt.ru/forum/search.php?author_id=641&sr=topics&filter=topics_answered
As you can see the number of answers is 32 but only 4 are shown. Note: that there are two pages shown in the result. If you click page 2 then five more topics will be shown (but still not 32). So maybe the bug is related to the number of total topics the user has.

**Comment by Galixte** [comment]
Hi @kinerity,

could you take a look, please ?

**Comment by MyLifeOrganized** [comment]
More details. If I run this sql for the required <%author_id%> I do get the correct number of posts as result:

```
SELECT t.topic_id, topic_title FROM phpbb_topics t 
  LEFT JOIN phpbb_posts p ON (p.post_id = t.bestanswer_id)
WHERE 
p.poster_id = <%author_id%>;
```

**Comment by MyLifeOrganized** [comment]
Hi @kinerity,
 
The same problem in v.0.0.1.

Please let me know if you need more details on this issue.

Thanks!

**Comment by kinerity** [comment]
You'll get quicker responses in the development topic. I'll look into this.

**Comment by kinerity** [comment]
Hmm, not sure exactly what to tell you. I just tested this on my test board (2 forums, 7 topics, 29 posts) and the extension works as intended. Can you enable `DEBUG` and `DEBUG_CONTAINER` and post back when it's enabled? That will allow me to see the SQL queries being executed.

**Comment by MyLifeOrganized** [comment]
I have enabled debug mode on test server and got these two SQL queries where bestanswer was mentioned:

```
SELECT p.post_id, p.poster_id, t.topic_id, t.bestanswer_id FROM (phpbb_posts p CROSS JOIN phpbb_topics t) WHERE p.post_id = t.bestanswer_id
AND p.poster_id = 2286
```


```
SELECT t.*, f.forum_id, f.forum_name, tp.topic_posted, tt.mark_time, ft.mark_time as f_mark_time, p.post_id, p.poster_id
FROM phpbb_topics t
LEFT JOIN phpbb_forums f ON (f.forum_id = t.forum_id)
LEFT JOIN phpbb_topics_posted tp ON (tp.user_id = 2
AND t.topic_id = tp.topic_id) LEFT JOIN phpbb_topics_track tt ON (tt.user_id = 2
AND t.topic_id = tt.topic_id)
LEFT JOIN phpbb_forums_track ft ON (ft.user_id = 2
AND ft.forum_id = f.forum_id) LEFT JOIN phpbb_posts p ON (p.post_id = t.bestanswer_id)
WHERE t.topic_id IN (3509, 3596, 3175, 3588, 3595, 3594, 3591, 3587, 3579, 3586, 3583, 3584, 3585, 3580, 3514, 3578, 3577, 3087, 3570, 3548, 3530, 3517, 1268, 2848, 3511) AND (t.topic_visibility = 1 OR t.forum_id IN (8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 43, 44, 45, 46, 47)) AND p.poster_id = 2286
ORDER BY t.topic_last_post_time DESC
```

I am not sure if it is correct that many forum/topic id mentioned in IN ().

**Comment by kinerity** [comment]
That still looks correct? Have you tried this on a clean install with no other extensions?

**Comment by MyLifeOrganized** [comment]
I tried to disable (not uninstall) all extensions on the test mirror of the server and the result was the same.
The problem is that the issue is reproduced on other forum (I do not  even know the owner of it): https://dmyt.ru 
So it is definitely not the issue of my forum only. We just need to understand the reason.

Please let me know what other information I can provide to help with this issue. This extension is critical for our forum. Thanks for implementing it!

**Comment by MyLifeOrganized** [comment]
I just executed both SQLs above against the db to check the results. The first one produced 7 records as expected. The second 0 records. The second SQL was generated in debug mode when I was logged in to the forum and as I said in this case the test URL produces 0 records instead of 3 when you do not login.

**Comment by MyLifeOrganized** [comment]
After detailed investigation I have additional info for the issue:

1) I have removed this part from the SQL:

`t.topic_id IN (3509, 3596, 3175, 3588, 3595, 3594, 3591, 3587, 3579, 3586, 3583, 3584, 3585, 3580, 3514, 3578, 3577, 3087, 3570, 3548, 3530, 3517, 1268, 2848, 3511)`

So the resulted SQL is like this:

```
SELECT t.*, f.forum_id, f.forum_name, tp.topic_posted, tt.mark_time, ft.mark_time AS f_mark_time, p.post_id, p.poster_id
FROM phpbb_topics t
LEFT JOIN phpbb_forums f ON (f.forum_id = t.forum_id)
LEFT JOIN phpbb_topics_posted tp ON (tp.user_id = 2 AND t.topic_id = tp.topic_id) 
LEFT JOIN phpbb_topics_track tt ON (tt.user_id = 2 AND t.topic_id = tt.topic_id)
LEFT JOIN phpbb_forums_track ft ON (ft.user_id = 2 AND ft.forum_id = f.forum_id) 
LEFT JOIN phpbb_posts p ON (p.post_id = t.bestanswer_id)
WHERE p.poster_id = 2286
AND (t.topic_visibility = 1 OR t.forum_id IN (8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 43, 44, 45, 46, 47))

ORDER BY t.topic_last_post_time DESC
```

As a result I got all 7 topics the user answered - the correct expected list (IDs):
```
3533
3519
3512
3222
2527
3302
2808
```

2) As you can see these IDs are not in the initial generated SQL for t.topic_id IN (). Why are they needed? What part of the forum constructed this part of the SQL: the phpBB or the extension? It is obvious that this list of IDs in the "IN(...)" is the reason why the correct list of topics could not be generated.

3) I have an idea that this maybe related to the pages in the result. Maybe when you have many topics divided by pages in the result the extension does something wrong or does not consider this?

4) The other idea is permissions. This user (2286) belongs to a group which could access more forums than regular user. But all the answered topics are in open forums.

**Comment by kinerity** [comment]
I will look into this further tonight. I'm guessing it has to do with something in the way the SQL queries are built into the search page. I'm currently at work and can't work on it.

**Comment by kinerity** [comment]
> t.topic_id IN
OK, I at least know where this is coming from now, but unfortunately, it's not something I can fix. It's coming directly from phpBB and it's part of the core search code. I'm still unable to duplicate the issue you're seeing. If I search for topics answered and the list of topics is more then to topics per page setting, it correctly adds pagination, but will show the total as the correct amount. To test this, I set the topics per page to 1 and had 2 topics answered. You, however, are experiencing something I still can't wrap my head around.

Regardless, I'll see if there is another way to list the topics answered. It will be a while, so please bear with me.

**Comment by kinerity** [comment]
OK, download this -> http://www.acsyste.com/bestanswer.zip

Replace the services.yml and listener.php files with the ones in the download above. Clear your cache file by deleting everything except for .htaccess and index.htm

Tell me if it resolves your issue. **YES, I KNOW AOBUT THE "Undefined index" DEBUG ERROR**

**Comment by MyLifeOrganized** [comment]
@kinerity the url you sent to me does not work (returned 404).  I am not sure how to contact your directly from GitHub. I am on vacation starting from 10/23 so will be able to try the fix later. Thanks!

**Comment by kinerity** [comment]
Yeah, the link was removed after it was discovered to contain a serious bug. I'm still working on it.

**Comment by kinerity** [comment]
I honestly don't know what to say here. I'm still unable to duplicate the issue you're running into (although it is valid). Also, to rewrite the search portion of the extension will require some extensive changes (a new controller, reworking the routes, as well as implementing the same code that's already in search.php) and I don't see that as a viable solution - at least not until a 2nd version of this extension. I'll keep trying, but without a solid set of parameters that cause this issue to appear, it's hard to implement a proper solution.
